### PR TITLE
Create function to join on misaligned dates

### PR DESCRIPTION
**Pull request by MarionJHolloway**
# Description
https://trello.com/c/eTLzu2PQ/129-epic-create-a-generic-date-matrix-function-for-selecting-which-importdates-should-be-used-to-join-datasets-together-must

## Functionality included
* Aligns import date columns from two dataframes
* Joins two dataframes with mismatching import dates on their import dates and one additional column
* Reduces the size of the secondary dataset to minimise the size of the cross join involved

I expect we will need to iterate on this as the pipeline takes shape. Feedback on how to make this easy to improve on in future tickets would be helpful.

# How to test
Unit tests passing
Run in branch successfully with test data

# Developer checklist
- [x] Unit test
- [x] Linked to Trello ticket
- [ ] Documentation up to date

### CODE REVIEW

File: utils/cleaning_utils.py [hunk 1]
```diff
@@ -119,3 +120,104 @@ def column_to_date(
...
+) -> DataFrame:
+    primary_dates = primary_df.select(primary_column).dropDuplicates()
+    secondary_dates = secondary_df.select(secondary_column).dropDuplicates()
+    min_secondary_date = calculate_min_secondary_date(
+        primary_dates, secondary_dates, primary_column, secondary_column
+    )
+    if min_secondary_date != None:
+        secondary_dates = secondary_dates.where(
+            secondary_dates[secondary_column] >= F.lit(min_secondary_date)
+        )
```
Comment (james-francis-MT) [inline utils/cleaning_utils.py:156]: Nice optimisation

File: utils/column_names/cleaned_data_files/ascwds_workplace_data_columns.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from dataclasses import dataclass
+
+from utils.column_names.raw_data_files.ascwds_workplace_columns import (
+    AscwdsWorkplaceColumns,
+)
+
+
+@dataclass
+class AscwdsWorkplaceCleanedColumns(AscwdsWorkplaceColumns):
+    cleaned_import_date: str = "ascwds_workplace_import_date"
```
Comment (james-francis-MT) [inline utils/column_names/cleaned_data_files/ascwds_workplace_data_columns.py:10]: Think this file is a duplication of `utils/column_names/cleaned_data_files/ascwds_workplace_cleaned_values.py`

**Review by james-francis-MT — APPROVED** [review]
Nice work, it would be good to get this tested asap in an actual cleaning script

File: utils/column_names/cleaned_data_files/ascwds_workplace_data_columns.py (hunk 2 above)
Comment (MarionJHolloway) [reply utils/column_names/cleaned_data_files/ascwds_workplace_data_columns.py:10]: Good spot, I think that's because I branched a week ago!

**Pull request merged**